aws_advanced_python_wrapper/limitless_connection_plugin.py

```diff
@@ -0,0 +1,543 @@
+        return self._props
+
+    def get_connection(self):
+        return self._connection
+
+    def set_connection(self, connection: Connection):
+        if self._connection is not None and self._connection != connection:
+            try:
+                self._connection.close()
+            except Exception:
+                pass
+
+        self._connection = connection
+
+    def get_connect_func(self) -> Callable:
+        return self._connect_func
+
+    def get_limitless_routers(self):
+        return self._limitless_routers
+
+    def set_limitless_routers(self, limitless_routers: List[HostInfo]):
+        self._limitless_routers = limitless_routers
+
+    def get_connection_plugin(self):
+        return self._connection_plugin
+
+    def is_any_routers_available(self):
+        for router in self.get_limitless_routers():
+            if router.get_availability() == HostAvailability.AVAILABLE:
+                return True
+        return False
+
+
+class LimitlessRouterService:
+    _CACHE_CLEANUP_NS: int = 6 * 10 ^ 10  # 1 minute
+    _limitless_router_cache: ClassVar[SlidingExpirationCacheWithCleanupThread[str, List[HostInfo]]] = \
+        SlidingExpirationCacheWithCleanupThread(_CACHE_CLEANUP_NS,
+                                                should_dispose_func=lambda x: True,
+                                                item_disposal_func=lambda x: {})
+
+    _limitless_router_monitor: ClassVar[SlidingExpirationCacheWithCleanupThread[str, LimitlessRouterMonitor]] = \
+        SlidingExpirationCacheWithCleanupThread(_CACHE_CLEANUP_NS,
+                                                should_dispose_func=lambda monitor: True,
+                                                item_disposal_func=lambda monitor: monitor.close())
+
+    def __init__(self, plugin_service: PluginService, create_limitless_router_monitor: Callable, query_helper: LimitlessQueryHelper):
+        self._plugin_service = plugin_service
+        self._create_limitless_router_monitor = create_limitless_router_monitor
+        self._query_helper = query_helper
+
+        self._force_get_limitless_routers_lock_map: ConcurrentDict[str, RLock] = ConcurrentDict()
+        self._rds_utils = RdsUtils()
+
+    def establish_connection(self, context: LimitlessConnectionContext) -> None:
+        context.set_limitless_routers(self._get_limitless_routers(context.get_host_info().host_id, context.get_props()))
+
+        if context.get_limitless_routers() is None or len(context.get_limitless_routers()) == 0:
+            logger.debug("LimitlessRouterServiceImpl.limitlessRouterCacheEmpty")
+
+            wait_for_router_info = WrapperProperties.WAIT_FOR_ROUTER_INFO.get(context.get_props())
+            if wait_for_router_info:
+                self._synchronously_get_limitless_routers_with_retry(context)
+            else:
+                logger.debug("LimitlessRouterServiceImpl.UsingProvidedConnectUrl")
+                if context.get_connection() is None:
```

Comment:
In JDBC we also check if the connection is closed, I'm not sure if this is necessary or not though